src/shadowbox/server/shared_metrics.ts

```diff
@@ -12,192 +12,150 @@
-      }
-      getHourlyServerMetricsReport(
-          this.serverConfig.data().serverId, startDatetime, endDatetime, lastHourUserStats,
-          ipLocationService)
-          .then((report) => {
-            if (report) {
-              postHourlyServerMetricsReports(report, metricsUrl);
-            }
-          });
-    });
+// Holds one hour usage metrics in memory.
+// TODO: migrate to an implementation that uses Prometheus.
+export class InMemoryOneHourUsageMetrics implements OneHourUsageMetrics, UsageMetricsRecorder {
+  // Map from the metrics AccessKeyId to metrics (bytes transferred, IP addresses).
+  private lastHourUsage = new Map<AccessKeyId, KeyUsage>();
 
-    // Set hourly metrics report interval
-    setHourlyInterval(this.generateHourlyReport.bind(this));
+  getOneHourUsage(): KeyUsage[] {
+    return [...this.lastHourUsage.values()];
   }
 
-  // CONSIDER: accepting hashedIpAddresses, which can be persisted to disk
-  // and reported to the metrics server (to approximate number of devices per userId).
-  recordBytesTransferred(userId: AccessKeyId, numBytes: number, ipAddresses: string[]) {
-    const perUserMetrics = this.lastHourUserMetrics.get(userId) ||
-        {bytesTransferred: 0, anonymizedIpAddresses: new Set<string>()};
-    perUserMetrics.bytesTransferred += numBytes;
-    const anonymizedIpAddresses = getAnonymizedAndDedupedIpAddresses(ipAddresses);
-    for (const ip of anonymizedIpAddresses) {
-      perUserMetrics.anonymizedIpAddresses.add(ip);
+  // We use a separate metrics id so the accessKey id is not disclosed.
+  recordBytesTransferred(accessKeyMetricsId: AccessKeyId, numBytes: number, countries: string[]) {
+    // Don't record data for sanctioned countries.
+    for (const country of countries) {
+      if (SANCTIONED_COUNTRIES.has(country)) {
+        return;
+      }
+    }
+    if (numBytes === 0) {
+      return;
+    }
+    let keyUsage = this.lastHourUsage.get(accessKeyMetricsId);
+    if (!keyUsage) {
+      keyUsage = {accessKeyMetricsId, inboundBytes: 0, countries: new Set<string>()};
+      this.lastHourUsage.set(accessKeyMetricsId, keyUsage);
+    }
+    keyUsage.inboundBytes += numBytes;
+    for (const country of countries) {
+      keyUsage.countries.add(country);
     }
-    this.lastHourUserMetrics.set(userId, perUserMetrics);
-    this.toJson(this.config.data());
-    this.config.write();
   }
+}
 
-  reset(): void {
-    this.lastHourUserMetrics = new Map<AccessKeyId, PerUserMetrics>();
-    this.startDatetime = new Date();
-    this.toJson(this.config.data());
-    this.config.write();
-  }
+// Keeps track of the connection metrics per user, since the startDatetime.
+// This is reported to the Outline team if the admin opts-in.
+export class OutlineSharedMetricsReporter implements SharedMetricsReporter {
+  // Date+time at which we started recording connection metrics, e.g.
+  // in case this object is constructed from data written to disk.
+  private startDatetime: Date;
```

Comment:
I always thought this was an unfortunate name. It's really the time at which metrics were last reported, right?